cumulusci/salesforce_api/org_schema.py

```diff
@@ -0,0 +1,312 @@
+            if resp.status == 200
+        ]
+        unexpected = [resp for resp in responses if resp.status not in (200, 304)]
+        for unknown in unexpected:
+            logger.warn(
+                f"Unexpected describe reply. An SObject may be missing: {unknown}"
+            )
+        self._populate_cache_from_describe(changes, last_modified_date)
+
+    def _populate_cache_from_describe(
+        self, describe_objs: List[Tuple[dict, str]], last_modified_date
+    ):
+        """Populate a schema cache from a list of describe objects."""
+        engine = self.engine
+        metadata = Base.metadata
+        metadata.bind = engine
+        metadata.reflect()
+
+        with BufferedSession(engine, metadata) as sess:
+
+            max_last_modified = (parsedate(last_modified_date), last_modified_date)
+            for (sobj_data, last_modified) in describe_objs:
+                fields = sobj_data.pop("fields")
+                create_row(sess, SObject, sobj_data)
+                for field in fields:
+                    field["sobject"] = sobj_data["name"]
+                    create_row(sess, Field, field)
+                    sortable = parsedate(last_modified), last_modified
+                    if sortable > max_last_modified:
+                        max_last_modified = sortable
+
+            create_row(
+                sess,
+                FileMetadata,
+                {"name": "Last-Modified", "value": max_last_modified[1]},
+            )
+            create_row(sess, FileMetadata, {"name": "FormatVersion", "value": 1})
+
+        engine.execute("vacuum")
+
+
+def create_row(buffered_session: "BufferedSession", model, valuesdict: dict):
+    buffered_session.write_single_row(model.__tablename__, valuesdict)
+
+
+class BufferedSession:
+    """Buffer writes to a SQL DB for faster performmance"""
+
+    def __init__(self, engine: Engine, metadata: MetaData, max_buffer_size: int = 1000):
+        self.buffered_rows = defaultdict(list)
+        self.columns = {}
+        self.engine = engine
+        self.metadata = metadata
+        self._prepare()
+        self.max_buffer_size = max_buffer_size
+
+    def __enter__(self, *args):
+        self.session = create_session(bind=self.engine, autocommit=False)
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+    def _prepare(self):
+        # Setup table info used by the write-buffering infrastructure
+        assert self.metadata.tables
+        self.insert_statements = {}
+        for tablename, model in self.metadata.tables.items():
+            self.insert_statements[tablename] = model.insert(bind=self.engine)
+            self.columns[tablename] = {
+                colname: None for colname in model.columns.keys()
+            }
+
+    def write_single_row(self, tablename: str, row: Dict) -> None:
+        # but first, normalize it so all keys have a value. SQLite Requires it.
+        row = {**self.columns[tablename], **row}
```

Comment:
What happens in a future API version when there are additional properties included in the describe result that aren't part of our table schema?

Reply:
Experimentally, adding new keys does not seem to phase SQL Alchemy. The model knows what fields it is expecting so it is not surprising that it "looks or" those fields to binds and just ignores the rest, but I couldn't find this quirk documented anywhere, so I reimplemented in case the semantics change.